### PR TITLE
Data Processing and Model Training of Tox21-NR-AR Dataset- Arpita Kesharwani

### PR DESCRIPTION
**Pull request by KesharwaniArpita**
## 🏆 Work Done This Week  
Over the past week, I have:  
- Explored the **Tox21 dataset** and documented its significance for computational toxicology.  
- Developed Python scripts for **data loading (`dataloader.py`)** and **exploratory data analysis (`eda.py`)**, ensuring seamless dataset handling.  
- Created Jupyter notebooks (`tox21_data_exploration_all.ipynb` and `tox21_data_exploration_single.ipynb`) to analyze toxicity endpoints and facilitate **data-driven insights**.  
- Implemented a structured **EDA pipeline**, including summary statistics, missing values analysis, feature distribution plots, and outlier detection.  
- Documented the workflow to ensure easy reproducibility for Outreachy contributors.  

This work lays the foundation for further featurisation and ML modeling on the data! 🚀  

---

This PR adds:  
✅ Python scripts for **loading and preprocessing Tox21 data** (`dataloader.py`)  
✅ A **complete exploratory data analysis (EDA) pipeline** (`eda.py`)  
✅ Two **Jupyter notebooks** for dataset exploration (`tox21_data_exploration_all.ipynb` & `tox21_data_exploration_single.ipynb`)  
✅ **Detailed documentation** on dataset usage, EDA insights, and workflow reproducibility  

This contribution aims to **streamline dataset analysis and prepare for ML modeling**. Feedback is welcome! 😊  

---

### CODE REVIEW

**Comment by KesharwaniArpita** [comment]
@GemmaTuron, can you kindly review my PR?

**Comment by KesharwaniArpita** [comment]
Hi @GemmaTuron, I have completed the featurisation of the dataset using Erisila, can you provide your review on the same?

**Comment by KesharwaniArpita** [comment]
This time the changes integrate Ersilia’s pre-trained compound embeddings for molecular featurization. The selected model generates vector representations for SMILES data, which are used to improve toxicity prediction on Tox21.

**Why Compound Embeddings?**  
Compound embeddings provides **richer molecular representations** by learning from large datasets, making them more generalizable than traditional featurizers (like Morgan fingerprints). Unlike **task-specific models** (e.g., cardiotoxicity predictors), embeddings can be used for **multiple toxicity endpoints**, making them **flexible, transferable, and often more informative** across Tox21 tasks. 🚀

#### **Next Steps**  
- Use featurized data from Ersilia embeddings to train a **Neural network** for all 12 Tox21 tasks.  
- Implement output for all the 12 tasks with `binary_crossentropy` loss and `sigmoid` activation.  
- Train and evaluate the model, comparing performances.  
- Optimize hyperparameters and assess results across tasks.

**Comment by KesharwaniArpita** [comment]
My recent contribution include: data preprocessing and model training for the NR-AR assay in the Tox21 dataset. The updates are described below:

## Data Preprocessing Summary

I have developed a deep learning model to predict androgen receptor (AR) activity from molecular structures, addressing severe class imbalance (4.25% positives) in the TOX21 NR-AR dataset. For a smooth model training it is utterly important to address the present data, that is why we first did feaurisation and the same reason is we are next tackling class imbalance and performing the train, val, and test split(stratified).

## Dataset Overview
- **Total compounds**: 7,265 (6,956 negatives, 309 positives)
- **Severe class imbalance**: 4.25% positive samples
- **Features**: 1024D molecular embeddings + binary AR activity label

## Key Processing Steps

### 1. Feature Engineering: Expanded embeddings into 1024 numerical features

### 2. Data Splitting:
- **Stratified 80/10/10 split** (random_state=42)
  - Training: 5,812 compounds 
  - Validation: 727 compounds
  - Test: 727 compounds
- Maintained original imbalance in val/test sets

### 3. Class Imbalance Handling
- Applied **SMOTE only to training data**:
  - Generated synthetic positives
  - Final training distribution: 1:1 (5,812 each)
- Preserved real-world imbalance in val/test (22.5:1)

### 4. Feature Standardization
```python
scaler = StandardScaler()
X_train = scaler.fit_transform(X_resampled)  # Fit only on train
X_val = scaler.transform(val_df.filter(like="emb_"))
```

## Quality Control
- **Reproducibility**: All splits/seeding with random_state=42
- **No data leakage**: 
  - SMOTE only on training
  - Scaler fit only on training
- **Visual validation**:
  - Before SMOTE: 95.75% negative
  - After SMOTE: 50% positive
 

Class Distribution

After SMOTE


## Key Statistics
| Stage          | Negatives | Positives | Ratio |
|----------------|-----------|-----------|-------|
| Raw Data       | 6,956     | 309       | 22.5:1|
| SMOTE Training | 5,812     | 5,812     | 1:1   |
| Validation     | 696       | 31        | 22.5:1|
| Test           | 696       | 31        | 22.5:1|


This helps to emphasizes anti-leakage measures for model training.

## Model Training
After data preprocessing, we have obtained train, validation and test data. The model training was done on the train and validation data primarily.

### Model Architecture
```python
Sequential(
    Input(shape=(256,)),
    Dense(512, activation='relu') → BatchNorm → Dropout(0.3),
    Dense(256, activation='relu') → BatchNorm → Dropout(0.3), 
    Dense(128, activation='relu') → BatchNorm → Dropout(0.3),
    Dense(64, activation='relu') → BatchNorm → Dropout(0.3),
    Dense(1, activation='sigmoid')
)
```

### Key Design Choices
- **Progressive dimensionality reduction** (512→64)
- **ReLU activation** for hidden layers
- **Sigmoid output** for binary classification
- **Regularization**: BatchNorm + 30% Dropout
- **Optimizer**: Adam (lr=1e-5)

## Hyperparameters
| Parameter        | Value  | Purpose |
|------------------|--------|---------|
| Learning Rate    | 1e-5   | Stable gradient updates |
| Batch Size       | 1024   | Efficient GPU utilization |
| Epochs           | 150    | Full convergence |
| Dropout Rate     | 0.3    | Prevent overfitting |
| Hidden Layers    | 4      | Optimal complexity |

## Performance Results
| Dataset    | Accuracy | F1-Score | Recall | Precision |
|------------|----------|----------|--------|-----------|
| Training   | 0.98     | 0.98     | 0.98   | 0.98      |
| Validation | 0.93     | 0.94     | 0.94   | 0.94      |
| Test       | 0.90     | 0.93     | 0.91   | 0.95      |

Training History

ROC curve

<p align="center">
  Image 1
  Image 2
</p>


## Key Insights
- Achieved **93% F1-score** on imbalanced test set
- **91% recall** demonstrates effective minority class identification
- Training curves show no overfitting (val loss plateaus)
- t-SNE visualization confirmed:
  - Clear separation of active/inactive compounds
  - SMOTE generated chemically meaningful synthetic samples

> "The model demonstrates robust performance while maintaining generalizability to real-world imbalanced data."\

## Model Testing
Model testing was done on test data (Tox21-NR-AR dataset) and Tox21-NR-AR-LBD dataset. The results of both of them are summarised below. The best model has come out to be `3_tox21_classifier.h5`.

### NR-AR Test Set Results Summary 

| **Metric**             | **Value** | **Interpretation**                                 |
|------------------------|----------:|----------------------------------------------------|
| **Accuracy**           | 91.0%     | High overall prediction rate.                      |
| **Class 0 Precision**  | 98%       | Most predicted non-events are correct.             |
| **Class 0 Recall**     | 92%       | Model captures most of the true non-events.        |
| **Class 1 Precision**  | 26%       | High false positive rate for the minority class.   |
| **Class 1 Recall**     | 61%       | Detects 61% of actual positive (minority class).   |
| **Macro F1-Score**     | 0.66      | Balanced metric across both classes.               |
| **Weighted F1-Score**  | 0.93      | Weighted average considers class imbalance.        |

#### 📊 **Class Distribution**
- **Class 0.0 (Negative)**: 95.7% (696 samples)  
- **Class 1.0 (Positive)**: 4.3% (31 samples)  

#### 🔁 **Confusion Matrix**

|                     | **Predicted: Class 0** | **Predicted: Class 1** |
|---------------------|------------------------|------------------------|
| **Actual: Class 0** | 641                    | 55                     |
| **Actual: Class 1** | 12                     | 19                     |

<p align="center">
  Image 1
  Image 2
</p>

The model shows high overall accuracy (91%) with excellent performance on the majority class (0.0). However, it struggles a little with the minority class (1.0), despite a moderate recall (61%). This indicates a **class imbalance issue**, where the model favors non-events and misclassifies many positives.

#### NR-AR-LBD Results Summary 

| **Metric**          | **Value**  | **Interpretation** |
|---------------------|------------|--------------------|
| **Accuracy**        | 81.93%     | Decent overall prediction rate. |
| **ROC AUC**         | 0.826      | Good class separation ability. |
| **Positive Recall** | 70%        | Detects **70% of toxic compounds** correctly. |
| **Positive Precision** | 13%  | **High false positives** due to class imbalance. |

#### 📊Class Distribution 
- **Non-Toxic (0.0)**: 96.5% (6,521 samples)  
- **Toxic (1.0)**: 3.5% (237 samples)  

#### 🔁 Confusion Matrix  
|                     | Predicted Non-Toxic | Predicted Toxic |
|---------------------|---------------------|-----------------|
| **Actual Non-Toxic**| 5,371               | 1,150           |
| **Actual Toxic**    | 71                  | 166             |  

#### ROC Curve and TSNE of Predicted Outcomes
<p align="center">
  Image 1
  Image 2
</p>

✅ **Model generalizes reasonably well** (AUC = 0.826) despite being trained on a different endpoint.  
⚠ **Severe class imbalance** affects precision (only 13% of predicted toxic compounds are correct).  
🔍 **t-SNE plots** show some clustering of toxic compounds, but overlap exists. 


---

### **Areas for Refinement**  
- **Test Set Generalization**: Slight AUC drop suggests tuning for real-world variability (e.g., augment data diversity).  
- **False Positives**: Confusion matrices show ~12 FP in test set – consider cost-sensitive learning if false negatives are critical.
